### ISSUE TITLE
Recursively broadcasting does not work

### ISSUE THREAD

**Issue by adagioburner** [jmoenig/Snap--Build-Your-Own-Blocks#955]
Broadcasting after receiving a message yields unpredictable results.

http://snap.berkeley.edu/snapsource/snap.html#present:Username=adagioburner&ProjectName=123

In this project, sprite 1 broadcasts message "one" once, and sprite 2 broadcasts "two" after receiving a message from 1. So I would expect 3 to get one message from each, In reality, sprite 3 gets *two* messages from 2 and none from 1, which sounds wrong to me :(

**Comment by adagioburner** [comment]
Note that works in Scratch (although you of course need to write more code to handle each message type separately)
https://scratch.mit.edu/projects/78895488/

**Comment by brianharvey** [comment]
You are imagining that the three sprites truly run in parallel, but of course only one script is actually running at any moment.  So, sprite 1 broadcasts "one," and then its turn ends and sprite 2 gets the message.  It broadcasts "two," then it's sprite 3's turn and it gets the "two" -- the most recent message.  Then it's sprite 1's turn again, and it gets the "two" message, but doesn't broadcast anything.  And so on.

You'll get slightly different behavior, still not quite what you want, if you choose "thread safe scripts" in the settings menu.  Normally, if a second relevant event happens while a script with that trigger is running, the first run is stopped, and the script restarts from the beginning.  If thread safe scripts is set, the second event is just ignored, so that the script always finishes with consistent state.

Yes, ideally, there would be an event queue and all the broadcasts would be received by all sprites with the correct message.  Instead of a global MESSAGE reporter, there'd have to be an upvar in the hat block, or something.  It's on our list for someday.

**Comment by adagioburner** [comment]
Brian, I do not think any of the reasoning above can explain how sprite 3 can get two copies of message "two". 

With "thread safe scripts" execution simply deadlocks for me, not exactly the desired behavior ;)

As I am saying, this behaves correctly in Scratch (sprite 3 received one message "one" and one message "two"). So maybe it's worth another look?

**Comment by brianharvey** [comment]
Your program structure is different in Scratch.  In Snap!, by the time sprite 3 gets a chance to run, the (single, global) MESSAGE has been set to "two".  As I said, having a single, global MESSAGE isn't adequate to what you're trying to do here.  I've been having the same problem with a TCP/IP simulator I'm working on.  But it's a "documented bug."  That is, it's not something that'll be fixed instantly.

**Comment by adagioburner** [comment]
Right, globals are evil :)
That's too bad... If one of the kids in my class does something like that I'll have a hard time explaining what went wrong. Well, thanks for the explanations! I'll wait for a future release.

**Comment by jmoenig** [comment]
No, global variables are not evil, they simply reflect shared resources, like the mouse, the keyboard, the stage, the tempo, the timer, and broadcast messages. If one script has a CLEAR command while another is drawing something there are similar conditions.

**Comment by adagioburner** [comment]
Hmm, I hardly feel that broadcast messages fit into that list. They are simply means of communication, and, if shared, they kind of suck at that.

With the axiom "if I broadcast x, everyone eventually receives x" gone, thinking about what can happen becomes very hard.

It means I can never use broadcasting in a block (who knows how the block is going to be used?) -- or I need to make sure everyone's aware of that block's implementation detail. 

More generally, if two people work on the same project, they should be very careful about using broadcasting (better stay away from it), or they will start canceling each other's messages.

In short, yeah, you can declare it a shared resource, but it becomes kind of useless after that.

That said, can you tell me if the following plan of action is feasible (I haven't really looked at Snap codebase yet, and I don't think I presently have the cycles, so this is at a very high level):

- make the "global message" a list, "messages"
- the message reporter returns the head of "messages"
- broadcast: if "messages" is not empty, add your message to it and return
- broadcast: when you finished broadcasting, remove the head of "messages". If "messages" is not empty, start broadcasting the new head.

**Comment by brianharvey** [comment]
Sure, that's what I meant about an event queue, kind of, but I'm not sure it's enough to have a global queue.  If two sprites have different-length (different numbers of yield points, to be precise) when-I-receive scripts, one of them will be ready for the next message before the other, so a global MESSAGE reporter still doesn't suffice.  It has to be a per-script queue, and a per-script MESSAGE.

**Comment by adagioburner** [comment]
I feel I need to spend some time reading the code before I can continue a meaningful conversation...

My assumption was that broadcast simply does broadcast-and-wait on a separate thread. If that's the case, then I guess you can deliver messages one by one, from one global queue -- you simply do not start message two until everyone finished processing message one. Otherwise I guess you'll need per-script queues?

**Comment by brianharvey** [comment]
I guess you could do that, but the user might not want the fast sprite to be slowed down by the slow sprite.  Arguably the really right thing would be reentrant scripts, so all sprites get all messages right away, but decoupling the process from the code will never fly with Jens.

**Comment by jmoenig** [comment]
RE:
>decoupling the process from the code will never fly with Jens

Oh, not at all! In fact, that's precisely what we're doing in GP. This "physical" identity of a script with its process is a trait of Scratch, designed to make simply beginners' projects better traceable, it's *clearly* not how one would design a programming language for CS students. But you're right that I wouldn't want to give up on Scratch's quasi physical execution model in Snap :-)

**Comment by brianharvey** [comment]
Yeah, that's what I meant.

**Comment by adagioburner** [comment]
As you can imagine, I have quickly implemented a message queue using Snap's own extensibility mechanisms:
![new-boradcast](https://cloud.githubusercontent.com/assets/14843681/10147871/5d82f570-65e5-11e5-87f1-de419d4c4c94.png)
Once you have created a new broadcast and assigned it to a global variable, you either run or launch it to send messages
![set-broadcast](https://cloud.githubusercontent.com/assets/14843681/10147965/cc592460-65e5-11e5-8d6b-533bec1c9db6.png)
![run-broadcast](https://cloud.githubusercontent.com/assets/14843681/10147968/d0d34016-65e5-11e5-898b-7de81a688912.png)
Of course it's a little awkward to use, but something like that is essential if you want to send a message after you received a message :) and make sure your message is delivered :)
Question: do you think it may be helpful to include a block like this into the standard blocks?

**Comment by brianharvey** [comment]
Generally speaking, if you can already do it in Snap!, that's a good reason not to add a primitive.  Anyway, I'm holding out for a version that doesn't make everyone else wait for each message to be fully processed by all sprites.

**Comment by jmoenig** [comment]
I *love* your message queue, it's beautiful, a global function with a closurized list. Nice!
